### PR TITLE
MooseSharedPointer -> std::shared_ptr in framework

### LINKED ISSUE #8501: Replace MooseSharedPointer with std::shared_ptr in Warehouses
### Description of the enhancement or error report

Replace MooseSharedPointer with std::shared_ptr in the Warehouses. Consider replacing all MooseSharedPointers throughout the framework for consistency.

### Rationale for the enhancement or information for reproducing the error

1. We should do this anyway since MooseSharedPointer is no longer necessary with c++11
2. I'm not happy with all the complicated logic I used in #8497 (i.e. Checking all the other execute_on flags before attempting to delete objects with a specific execute_on flag). This is necessary because every time we insert an object we also add it to the `_all_objects` list so to remove an object from this list we have to make sure it's not used elsewhere (for integrity). I believe a better design would be to use `std::weak_ptrs` in the `_all_objects` container. This would simplify a lot of logic. We could simply remove all "strong" pointers from an individual warehouse without thinking about whether or not the object was being pointed to by other execute_on containers. If we delete the last execute_on pointers, the objects will clean themselves up. 

### Identified impact
(i.e. Internal object changes, limited interface changes, public API change, or a list of specific applications impacted)

We'll have to be careful about access to the weak pointers in _all_containers because they could now potentially be invalid.

### PR DESCRIPTION
**Pull request by permcody**
Getting rid of MooseSharedPointer everywhere in the framework and modernizing affected lines.

closes #8501

### CODE REVIEW

**Review by jwpeterson — APPROVED** [review]

File: framework/include/utils/MooseTypes.h [hunk 1]
```diff
@@ -27,16 +27,11 @@
...
-#  include <memory>
-#  define MooseSharedPointer std::shared_ptr
-#  define MooseSharedNamespace std
-#else
-#  include "boost/shared_ptr.hpp"
-#  define MooseSharedPointer boost::shared_ptr
-#  define MooseSharedNamespace boost
-#endif
+// DO NOT USE (Deprecated)
+#define MooseSharedPointer std::shared_ptr
```
Comment (aeslaughter) [inline framework/include/utils/MooseTypes.h]: Is it possible to make using MooseSharedPointer produce a warning?

Comment (permcody) [inline framework/include/utils/MooseTypes.h]: Not with a `#define`. It should be possible to promote this define to an actual shim class and then we could do that. MooseSharedPointer is still sprinkled throughout the modules a lot and also in applications so we probably don't want to do that just yet. However we'll have to be diligent with watching these sneak back into the framework while we are transitioning. It's a #define so there really isn't a problem if they do crop up from time to time.

Comment (aeslaughter) [inline framework/include/utils/MooseTypes.h]: Add it to the list of banned keywords in our pre-check.

Comment (permcody) [inline framework/include/utils/MooseTypes.h]: Too many instances still hanging around in modules plus this definition in MOOSE for that. The shim class will be better for supporting it with an error message for a little while, but not yet.

**Comment by jwpeterson** [comment]
> Too many instances still hanging around in modules plus this definition in MOOSE for that.

The banned keywords check runs on the entire codebase, and not just the diff, eh?  This also throws the idea for using that script to help upgrade MooseError out the window...

**Comment by permcody** [comment]
We could always work on making the pre-check smarter. There's no reason we can't have a diff precheck too. The MooseError issue is bigger than this since it's a different mechanism altogether, not just a #define.

**Comment by aeslaughter** [comment]
I thought it worked on the diff, oh well.

**Comment by permcody** [comment]
Any holdups? I have two more PRs that need these changes.

**Comment by jwpeterson** [comment]
I already approved it.  We can merge our own PRs if someone approves them, right?

**Pull request merged**